File: backend/src/routes/analytics.ts
import { Router } from 'express';
import { getDb } from '../db/db.js';
import { runFullPipeline } from '../pipelines/pipelineRunner.js';
import type { AnalyticsSummary } from '../types.js';

const router = Router();

// GET /api/analytics/summary - overall dashboard summary
router.get('/summary', (_req, res) => {
  const db = getDb();

  const totalPlants = (db.prepare('SELECT COUNT(*) as c FROM plants_raw').get() as any).c;
  const totalEvents = (db.prepare('SELECT COUNT(*) as c FROM care_events').get() as any).c;

  const avgMetrics = db.prepare(`
    SELECT AVG(health_score) as avg_health, AVG(care_compliance_rate) as avg_compliance
    FROM plant_health_metrics
    WHERE metric_date = (SELECT MAX(metric_date) FROM plant_health_metrics)
  `).get() as any;

  const mostActive = db.prepare(`
    SELECT plant_id, COUNT(*) as cnt FROM care_events
    WHERE scheduled_date >= datetime('now', '-30 days')
    GROUP BY plant_id ORDER BY cnt DESC LIMIT 1
  `).get() as any;

  const mostActivePlantName = mostActive
    ? (db.prepare('SELECT name FROM plants_raw WHERE id = ?').get(mostActive.plant_id) as any)?.name
    : null;

  const overdueCount = (db.prepare(`
    SELECT COUNT(DISTINCT p.id) as c FROM plants_raw p
    WHERE (
      SELECT MAX(scheduled_date) FROM events_raw WHERE plant_id = p.id AND type = 'water'
    ) < datetime('now', '-' || p.watering_interval || ' days')
  `).get() as any).c;

  const lastRun = (db.prepare(`
    SELECT finished_at FROM pipeline_runs WHERE status = 'success' ORDER BY id DESC LIMIT 1
  `).get() as any)?.finished_at;

  const summary: AnalyticsSummary = {
    totalPlants,
    totalEvents,
    avgHealthScore: Math.round(avgMetrics?.avg_health ?? 0),
    avgCompliance: Math.round((avgMetrics?.avg_compliance ?? 0) * 100),
    mostActivePlant: mostActivePlantName,
    overdueCount,
    lastPipelineRun: lastRun
  };

  res.json(summary);
});

// GET /api/analytics/plant/:id - per-plant analytics
router.get('/plant/:id', (req, res) => {
  const db = getDb();
  const plantId = req.params.id;

  const healthHistory = db.prepare(`
    SELECT metric_date, health_score, care_compliance_rate, total_events_30d
    FROM plant_health_metrics
    WHERE plant_id = ?
    ORDER BY metric_date DESC
    LIMIT 30
  `).all(plantId);

  const eventBreakdown = db.prepare(`
    SELECT status, COUNT(*) as count
    FROM care_events WHERE plant_id = ?
    GROUP BY status
  `).all(plantId);

  const feedbackBreakdown = db.prepare(`
    SELECT feedback, COUNT(*) as count
    FROM care_events
    WHERE plant_id = ? AND feedback IS NOT NULL
    GROUP BY feedback
  `).all(plantId);

  const avgDaysOverdue = (db.prepare(`
    SELECT AVG(ABS(days_overdue)) as avg FROM care_events
    WHERE plant_id = ? AND days_overdue IS NOT NULL
  `).get(plantId) as any)?.avg ?? 0;

  res.json({ healthHistory, eventBreakdown, feedbackBreakdown, avgDaysOverdue });
});

// GET /api/analytics/trends - time series trends
router.get('/trends', (_req, res) => {
  const db = getDb();

  // Weekly care events count
  const weeklyEvents = db.prepare(`
    SELECT
      strftime('%Y-W%W', scheduled_date) as week,
      COUNT(*) as total,
      SUM(CASE WHEN status = 'completed' THEN 1 ELSE 0 END) as completed,
      SUM(CASE WHEN status = 'skipped' THEN 1 ELSE 0 END) as skipped
    FROM care_events
    WHERE scheduled_date >= datetime('now', '-90 days')
    GROUP BY week
    ORDER BY week ASC
  `).all();

  // Health score trend across all plants
  const healthTrend = db.prepare(`
    SELECT metric_date, AVG(health_score) as avg_health, AVG(care_compliance_rate) as avg_compliance
    FROM plant_health_metrics
    WHERE metric_date >= date('now', '-30 days')
    GROUP BY metric_date
    ORDER BY metric_date ASC
  `).all();

  res.json({ weeklyEvents, healthTrend });
});

// GET /api/analytics/export - export all data as JSON
router.get('/export', (_req, res) => {
  const db = getDb();
  const plants = db.prepare('SELECT * FROM plants_raw').all();
  const events = db.prepare('SELECT * FROM care_events').all();
  const metrics = db.prepare('SELECT * FROM plant_health_metrics').all();
  const pipelineRuns = db.prepare('SELECT * FROM pipeline_runs ORDER BY id DESC LIMIT 10').all();

  res.setHeader('Content-Type', 'application/json');
  res.setHeader('Content-Disposition', 'attachment; filename="plantcare-data-export.json"');
  res.json({
    exportedAt: new Date().toISOString(),
    plants,
    events,
    metrics,
    pipelineRuns
  });
});

// POST /api/analytics/run-pipeline - manually trigger ETL
router.post('/run-pipeline', async (_req, res) => {
  try {
    const result = await runFullPipeline();
    res.json(result);
  } catch (e) {
    res.status(500).json({ error: (e as Error).message });
  }
});

// GET /api/analytics/pipeline-runs - pipeline audit log
router.get('/pipeline-runs', (_req, res) => {
  const db = getDb();
  const runs = db.prepare('SELECT * FROM pipeline_runs ORDER BY id DESC LIMIT 20').all();
  res.json(runs);
});

export default router;

File: backend/routes/analytics.py
from fastapi import APIRouter, HTTPException
from db import get_db
from pipelines import runner

router = APIRouter(prefix="/api/analytics")


@router.get("/summary")
def summary():
    try:
        with get_db() as conn:
            total_plants = conn.execute("SELECT COUNT(*) FROM plants_raw").fetchone()[0]
            total_events = conn.execute("SELECT COUNT(*) FROM care_events").fetchone()[0]
            completed = conn.execute(
                "SELECT COUNT(*) FROM care_events WHERE completed IS NOT NULL"
            ).fetchone()[0]
            overdue = conn.execute(
                "SELECT COUNT(*) FROM care_events WHERE days_overdue > 1"
            ).fetchone()[0]
            avg_health = conn.execute(
                "SELECT AVG(h.health_score) FROM plant_health_metrics h "
                "INNER JOIN (SELECT plant_id, MAX(computed_at) AS latest FROM plant_health_metrics GROUP BY plant_id) m "
                "ON h.plant_id = m.plant_id AND h.computed_at = m.latest"
            ).fetchone()[0]

        return {
            "total_plants": total_plants,
            "total_events": total_events,
            "completed_events": completed,
            "overdue_events": overdue,
            "compliance_rate": round(completed / total_events, 4) if total_events else 0,
            "avg_health_score": round(avg_health, 4) if avg_health else None,
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/trends")
def trends():
    try:
        with get_db() as conn:
            rows = conn.execute("""
                SELECT DATE(scheduled) AS day, COUNT(*) AS events,
                       SUM(CASE WHEN completed IS NOT NULL THEN 1 ELSE 0 END) AS completed
                FROM care_events
                GROUP BY day
                ORDER BY day DESC
                LIMIT 28
            """).fetchall()
        return [dict(r) for r in rows]
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/run-pipeline")
def trigger_pipeline():
    try:
        result = runner.run_pipeline()
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/export")
def export():
    try:
        with get_db() as conn:
            plants = [dict(r) for r in conn.execute("SELECT * FROM plants_raw").fetchall()]
            events = [dict(r) for r in conn.execute("SELECT * FROM care_events").fetchall()]
            metrics = [dict(r) for r in conn.execute("SELECT * FROM plant_health_metrics").fetchall()]
        return {"plants": plants, "events": events, "metrics": metrics}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/pipeline-runs")
def pipeline_runs():
    try:
        with get_db() as conn:
            rows = conn.execute(
                "SELECT * FROM pipeline_runs ORDER BY started_at DESC LIMIT 50"
            ).fetchall()
        return [dict(r) for r in rows]
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
